src/crystalline/broken_source_fixer.cr

```diff
@@ -0,0 +1,185 @@
+class Crystalline::BrokenSourceFixer
+  # Keep track of opening and closing keywords, and their idents,
+  # as they happen in the code.
+  record LineInfo,
+    line_index : Int32,
+    indent : Int32,
+    keyword : String
+
+  # Try to fix a broken source code by adding missing "end" and "}"
+  # according to indentation.
+  def self.fix(source : String) : String
+    # Keep a stack of opening keywords.
+    # We push to the stack when we find an opening keyword and
+    # we pop from the stack when we find a closing keyword,
+    # or when we find a wrong indentation.
+    stack = [] of LineInfo
+
+    lines = source.lines
+    lines.each_with_index do |line, line_index|
+      next if line.blank?
+
+      keyword = line_keyword(line)
+      indent = line_indent(line)
+
+      while true
+        last_info = stack.last?
+        break unless last_info
+
+        closing_keyword = closing_keyword(last_info)
+
+        # Nothing to fix unless there's a wrong indent
+        break unless wrong_indent?(indent, keyword, closing_keyword, last_info, line)
+
+        # We have a wrong indentation so we fix/close the opening keyword
+        # by adding an "end" (or "}") to it.
+        last_line = lines[line_index - 1]
+
+        lines[line_index - 1] =
+          if last_line.blank?
+            # If the line is empty we can change it to an end
+            # and even use the correct indent.
+            "#{("  " * last_info.indent)}#{closing_keyword}"
+          else
+            "#{last_line}; #{closing_keyword}"
+          end
+
+        stack.pop
+      end
+
+      # If we found an "end" at exactly the indentation of the last
+      # opening keyword, remove it from the stack.
+      if last_info && indent == last_info.indent && keyword == closing_keyword(last_info)
+        # all good: an end is closing an opening keyword
+        stack.pop
+        next
+      end
+
+      # Push to the stack if we found an opening keyword.
+      if keyword && !closing_keyword?(keyword)
+        stack << LineInfo.new(
+          line_index: line_index,
+          indent: indent,
+          keyword: keyword
+        )
+      end
+    end
+
+    while line_info = stack.pop?
+      lines[-1] = "#{lines[-1]}; #{closing_keyword(line_info)}"
+    end
+
+    lines.join("\n")
+  end
+
+  private def self.line_indent(line : String) : Int32?
+    non_whitespace_char_index = line.each_char_with_index do |char, i|
+      next if char.whitespace?
+      break i
+    end
+
+    if non_whitespace_char_index
+      non_whitespace_char_index // 2
+    else
+      0
+    end
+  end
+
+  private def self.line_keyword(line : String) : String?
+    if line.starts_with?(/\s*
```

Comment:
would it make sense to add `begin` to the list as well?

Reply:
Oh, good catch! I think I also missed macro, and rescue/ensure/else for begin and def. I'll add those on Monday.

Reply:
Done! I forgot to mention that Mondays happen on Wednesday in Argentina.

Reply:
> I forgot to mention that Mondays happen on Wednesday in Argentina.

Haha no worries in France we have the same concept 😄.

I'm not seeing any new commits though, did you forgot to push by any chance?

Reply:
Oh, I pushed to this repo instead of the fork. Now it's done.